executor/set.go

```diff
@@ -159,15 +160,23 @@ func (e *SetExecutor) setSysVariable(name string, v *expression.VarAssignment) e
 	}
 	newSnapshotTS := getSnapshotTSByName()
 	newSnapshotIsSet := newSnapshotTS > 0 && newSnapshotTS != oldSnapshotTS
-	// We don't check snapshot with gc safe point for read_ts
-	// Client-go will automatically check the snapshotTS with gc safe point. It's unnecessary to check gc safe point during set executor.
-	if newSnapshotIsSet && name != variable.TiDBTxnReadTS {
-		err = gcutil.ValidateSnapshot(e.ctx, newSnapshotTS)
-		if err != nil {
+	if newSnapshotIsSet {
+		if err = sessionctx.ValidateReadTS(ctx, e.ctx, newSnapshotTS); err != nil {
 			fallbackOldSnapshotTS()
 			return err
 		}
+
+		// We don't check snapshot with gc safe point for read_ts
+		// Client-go will automatically check the snapshotTS with gc safe point. It's unnecessary to check gc safe point during set executor.
+		if name != variable.TiDBTxnReadTS {
```

Comment:
There are other executors that need to be validated according to the [syntaxs](https://docs.pingcap.com/tidb/dev/as-of-timestamp#syntax), how about check it in client-go, the request must have `StaleRead` flag.

Reply:
If possible, I want to validate it when setting the timestamp. Otherwise, it looks counter-intuitive to report it until doing actual reads.

Reply:
FYI, the check of `start_ts < GC safe point` is also inside client-go